zschema/compounds.py

```diff
@@ -215,16 +226,26 @@ def to_dict(self):
         p = {self.key_to_es(k): v.to_dict() for k, v in source}
         return {"type":"subrecord", "subfields": p, "doc":self.doc, "required":self.required}
 
-    def validate(self, name, value):
-        if type(value) != dict:
-            raise DataValidationException("%s: %s is not a dict",
-                                          name, str(value))
+    def validate(self, name, value, policy=_NO_ARG, parent_policy=_NO_ARG):
+        calculated_policy = self._calculate_policy(name, policy, parent_policy)
+        try:
+            if not isinstance(value, dict):
+                raise DataValidationException("%s: %s is not a dict",
+                                              name, str(value))
+        except DataValidationException as e:
+            self._handle_validation(calculated_policy, e)
+            # cannot iterate over members if this isn't a dictionary
+            return
         for subkey, subvalue in sorted(value.iteritems()):
-            if not self.allow_unknown and subkey not in self.definition:
-                raise DataValidationException("%s: %s is not a valid subkey" %
-                                              (name, subkey))
-            else:
-                self.definition[subkey].validate(subkey, subvalue)
+            try:
+                if not self.allow_unknown and subkey not in self.definition:
+                    raise DataValidationException("%s: %s is not a valid subkey" %
+                                                  (name, subkey))
+                else:
+                    self.definition[subkey].validate(subkey, subvalue,
```

Comment:
This raises a `KeyError` if both `self.allow_unknown` and `subkey not in self.definition` (not a new problem); it seems the intent is `elif subkey in self.defnition`?

Reply:
I don't think that's right. The code is:
` if not self.allow_unknown and subkey not in self.definition:`

The key part being if **not** allow_unknown.

Reply:
ok, I see what you mean (although I think not what you wrote 😋). fixed.